plugins/search-backend-node/src/Scheduler.ts

```diff
@@ -14,52 +14,67 @@
  * limitations under the License.
  */
 
+import { AbortController } from 'node-abort-controller';
 import { Logger } from 'winston';
-import { runPeriodically } from './runPeriodically';
+import { TaskFunction, TaskRunner } from '@backstage/backend-tasks';
 
 type TaskEnvelope = {
-  task: Function;
-  interval: number;
+  task: TaskFunction;
+  scheduledRunner: TaskRunner;
 };
 
 /**
- * TODO: coordination, error handling
+ * @public ScheduleTaskParameters
  */
+export interface ScheduleTaskParameters {
+  id: string;
+  task: TaskFunction;
+  scheduledRunner: TaskRunner;
+}
 
 /**
  * @beta
  */
 export class Scheduler {
   private logger: Logger;
-  private schedule: TaskEnvelope[];
-  private runningTasks: Function[] = [];
+  private schedule: { [id: string]: TaskEnvelope };
+  private abortController: AbortController;
+  private isRunning: boolean;
 
   constructor({ logger }: { logger: Logger }) {
     this.logger = logger;
-    this.schedule = [];
+    this.schedule = {};
+    this.abortController = new AbortController();
+    this.isRunning = false;
   }
 
   /**
    * Adds each task and interval to the schedule.
    * When running the tasks, the scheduler waits at least for the time specified
    * in the interval once the task was completed, before running it again.
    */
-  addToSchedule(task: Function, interval: number) {
-    if (this.runningTasks.length) {
+  addToSchedule({ id, task, scheduledRunner }: ScheduleTaskParameters) {
+    if (this.isRunning) {
       throw new Error(
         'Cannot add task to schedule that has already been started.',
       );
     }
-    this.schedule.push({ task, interval });
+    this.schedule[id] = { task, scheduledRunner };
```

Comment:
Super minor: I wonder if it's worth checking if the key already exists and throwing instead of clobbering -- I guess previously we allowed multiple instances of a collator to be scheduled and run, though it's probably not a valid use-case.